osmaxx/profile/views.py

```diff
@@ -1,20 +1,141 @@
+        else:
+            self._ensure_profile_has_email(profile, user)
+        user.refresh_from_db()
+        return super().get(*args, **kwargs)
+
+    def is_new_user(self):
+        user = self.request.user
+        profile_is_new = user.profile.unverified_email is None
+        user_has_already_access = user.groups.filter(name=settings.OSMAXX_FRONTEND_USER_GROUP).exists()
+        return profile_is_new and not user_has_already_access
+
+    def post(self, *args, **kwargs):
+        response = super().post(*args, **kwargs)
+        if isinstance(response, HttpResponseRedirect):  # successful form validation
+            profile = Profile.objects.get(associated_user=self.request.user)
+            if profile.associated_user.email != profile.unverified_email:
+                self._send_email_verification(profile=profile)
+        return response
+
+    def _ensure_profile_has_email(self, profile, user):
+        if not profile.unverified_email:
+            profile.unverified_email = user.email
+            profile.save()
+
+    def _move_email_from_user_to_profile(self, user, profile):
+        if user.email:
+            profile.unverified_email = user.email
+            profile.save()
+            user.email = ''
+            user.save()
+
     def get_object(self, queryset=None):
-        return self.model.objects.get(pk=self.request.user.id)
+        profile, _ = Profile.objects.get_or_create(associated_user=self.request.user)
+        return profile
 
     def get_success_url(self):
-        messages.add_message(self.request, messages.SUCCESS, 'Profile successfully updated')
+        messages.add_message(self.request, messages.SUCCESS, _('Profile successfully updated'))
+        return reverse('profile:edit_view')
+
+
+class ActivationView(SendVerificationEmailMixin, LoginRequiredMixin, generic.UpdateView):
+    error_msg = _('Verification token too old or invalid. Please resend the verification email and try again.')
+
+    def get(self, request, *args, **kwargs):
+        user = request.user
+        token = request.GET.get('token', None)
+        if token:
+            data = user.profile.validate_key(activation_key=token)
+            if data:
+                user.email = data['email']
+                user.save()
+                self._set_group(user)
+                messages.add_message(self.request, messages.SUCCESS, _('Successfully verified your email address.'))
+            else:
+                messages.add_message(self.request, messages.ERROR, self.error_msg)
+        else:
+            messages.add_message(self.request, messages.ERROR, self.error_msg)
+        return redirect(reverse('profile:edit_view'))
+
+    def _set_group(self, user):
+        if settings.REGISTRATION_OPEN and not self._user_in_frontend_group(user):
+            self._grant_user_access(user)
+
+    def _user_in_frontend_group(self, user):
+        return user.groups.filter(name=settings.OSMAXX_FRONTEND_USER_GROUP).exists()
```

Comment:
Could this be turned into a (still private) module-level function, so that it could be used at [`ProfileView.is_new_user()`](https://github.com/geometalab/osmaxx/pull/670/files#diff-c95bcec7c46780ed79611fa9239b3ef1R71), too?